lte/cloud/go/sms_ll/sms_cp.go

```diff
@@ -84,29 +84,41 @@ func (cpm cpMessage) marshalBinary() []byte {
 }
 
 func (cpm *cpMessage) unmarshalBinary(input []byte) error {
-	// must be at least two octets long
-	if len(input) < 2 {
+	msgLen := len(input)
+	// Message length must be at least two octets long
+	if msgLen < 2 {
 		return smsCpError("Message too short")
 	}
 
 	cpm.firstOctet = input[0]
 	cpm.messageType = input[1]
 
-	switch cpm.messageType {
-	case CpData:
-		cpm.length = input[2]
-		cpm.rpdu = make([]byte, len(input[3:]))
-		copy(cpm.rpdu, input[3:])
-	case CpError:
-		if _, ok := CpCauseStr[input[2]]; ok {
-			cpm.cause = input[2]
+	if cpm.messageType != CpData && cpm.messageType != CpError && cpm.messageType != CpAck {
+		return smsCpError(fmt.Sprintf("Invalid Message type: %x", cpm.messageType))
+	}
+	// Decode IEs
+	decoded := 2
+
+	for decoded < msgLen {
```

Comment:
this is a weird for loop to me because it seems like if you ever execute the body more than once, you'll hit the error clause.

I would do something like

```
messageType := input[2]
switch messageType:
  case CpIei...:
  case: ...
  default:
    return error(...)
```

Reply:
there can be more then 1 IE packed.

Reply:
note how the decoded variable is incremented. If encoding is done right, after all IEs are checked, decoded will be msgLen. Check out the modified test case TestPiecewiseDecodeDeliveryFailure that packs both IEs in the message type CpError.

Reply:
and also TestPiecewiseDecodeDeliveryFailureChangeIEOrder that changes the IE orders.